joomla/joomla-cms
# PHP 8.1 compatibility

### Summary of Changes

Makes Joomla 4 compatible with PHP 8.1.

Tagging @zero-24 @wilsonge @bembelimen and @HLeithner — this impacts all Joomla versions you are leading the release of, not just 4.0, albeit this PR only targets 4.0 to preserve my and your sanity. 

Also tagging @rdeutz as the CMS maintenance liaison, @webgras as the CMS Release Team Leader and @tecpromotion as the CMS Release Team assistant leader — you guys know best if you need to bring this to the attention of other leadership members I may have missed.

### Testing Instructions

Install Joomla 4.0 on PHP 7.2, 7.3, 7.4 or 8.0

Create an article — you need to do that on PHP 7.2–8.0 because this is broken on PHP 8.1

Install PHP 8.1.0-RC6 on your server (Windows builds: https://github.com/shivammathur/php-builder-windows — macOS: see https://nunomaduro.com/how-to-install-php-8-1-rc-on-mac/)

Switch your site to use PHP 8.1.0-RC6

Try to edit the article you created

### Actual result BEFORE applying this Pull Request

The article title is empty, the article content is empty, you basically get the same interface as a blank article (except for tags and custom fields). Editing or creating an article is impossible. Trying to save changes doesn't work.

### Expected result AFTER applying this Pull Request

Editing the article works fine; the interface is populated. Any changes you make are saved fine. Creating an article also works without a problem.

### Documentation Changes Required

None

### Technical details

The problem happens due to the following change noted in [the official PHP 8.1 UPGRADING document](https://github.com/php/php-src/blob/PHP-8.1/UPGRADING):

```
  . When a method using static variables is inherited (but not overridden), the
    inherited method will now share static variables with the parent method.

        class A {
            public static function counter() {
                static $counter = 0;
                $counter++;
                return $counter;
            }
        }
        class B extends A {}

        var_dump(A::counter()); // int(1)
        var_dump(A::counter()); // int(2)
        var_dump(B::counter()); // int(3), previously int(1)
        var_dump(B::counter()); // int(4), previously int(2)

    This means that static variables in methods now behave the same way as
    static properties.
    RFC: https://wiki.php.net/rfc/static_variable_inheritance
```

The `\Joomla\CMS\Table\Table::getFields()` method used a method static variable to cache the fields of the table. However, since this method is inherited but not overridden in child table classes the static method variable was populated with the fields of whichever table class was loaded first — in the Joomla CMS this is always the`Joomla\CMS\Table\User` class for the `#__users` table. Since this information is used in the constructor to populate the missing table properties which are then used by the `bind()` method to decide which incoming source data to apply to the table object it became impossible to load any record or assign any data to any other table, causing this problem. It's actually surprising that only _so little_ broke; we got lucky!

I went through the entire core code with a fine comb and only found another three instances of method–level static variables which would cause a problem and I fixed them.

Yes, there are dozens more method–level static variables. However, these are either keyed to a globally unique key (either a database record ID or a unique ID constructed from immutable, globally unique data such as a component name) OR they were used to cache data which is immutable throughout the life of the request.

PS: CMSObject will need an attribute come PHP 8.2 and a rewrite come PHP 9.0. I'll get back to these next year. For now, it's important to make Joomla 4.0 ready for PHP 8.1 before it's released in two weeks time.
